RecoLocalTracker/SiPixelRecHits/src/PixelCPEClusterRepair.cc

```diff
@@ -0,0 +1,586 @@
+      "3D reconstruction failed with error " << theClusterParam.ierr2 << "\n";
+      
+      // GG: what do we do in this case?  For now, just return the cluster center of gravity in microns
+      // In the x case, apply a rough Lorentz drift average correction
+      // To do: call PixelCPEGeneric whenever PixelTempReco2D fails
+      float lorentz_drift = -999.9;
+      if ( ! GeomDetEnumerators::isEndcap(theDetParam.thePart) )
+         lorentz_drift = 60.0f; // in microns
+      else
+         lorentz_drift = 10.0f; // in microns
+      // GG: trk angles needed to correct for bows/kinks
+      if ( theClusterParam.with_track_angle )
+      {
+         theClusterParam.templXrec_ = theDetParam.theTopol->localX( theClusterParam.theCluster->x(), theClusterParam.loc_trk_pred ) - lorentz_drift * micronsToCm; // rough Lorentz drift correction
+         theClusterParam.templYrec_ = theDetParam.theTopol->localY( theClusterParam.theCluster->y(), theClusterParam.loc_trk_pred );
+      }
+      else
+      {
+         edm::LogError("PixelCPEClusterRepair")
+         << "@SUB = PixelCPEClusterRepair::localPosition"
+         << "Should never be here. PixelCPEClusterRepair should always be called with track angles. This is a bad error !!! ";
+         
+         theClusterParam.templXrec_ = theDetParam.theTopol->localX( theClusterParam.theCluster->x() ) - lorentz_drift * micronsToCm; // rough Lorentz drift correction
+         theClusterParam.templYrec_ = theDetParam.theTopol->localY( theClusterParam.theCluster->y() );
+      }
+   }   
+   else 
+   {
+      //--- Template Reco succeeded.
+      theClusterParam.hasFilledProb_ = true;
+
+      //--- Go from microns to centimeters
+      theClusterParam.templXrec_ *= micronsToCm;
+      theClusterParam.templYrec_ *= micronsToCm;
+      
+      //--- Go back to the module coordinate system
+      theClusterParam.templXrec_ += lp.x();
+      theClusterParam.templYrec_ += lp.y();
+   }
+   return;
+}
+
+
+
+
+//------------------------------------------------------------------
+//  localError() relies on localPosition() being called FIRST!!!
+//------------------------------------------------------------------
+LocalError
+PixelCPEClusterRepair::localError(DetParam const & theDetParam,  ClusterParam & theClusterParamBase) const
+{
+   
+   ClusterParamTemplate & theClusterParam = static_cast<ClusterParamTemplate &>(theClusterParamBase);
+   
+   //--- Default is the maximum error used for edge clusters.
+   //--- (never used, in fact: let comment it out, shut up the complains of the static analyzer, and save a few CPU cycles)
+   float xerr = 0.0f, yerr = 0.0f;
+   
+   // Check if the errors were already set at the clusters splitting level
+   if ( theClusterParam.theCluster->getSplitClusterErrorX() > 0.0f && theClusterParam.theCluster->getSplitClusterErrorX() < 7777.7f &&
+       theClusterParam.theCluster->getSplitClusterErrorY() > 0.0f && theClusterParam.theCluster->getSplitClusterErrorY() < 7777.7f )
```

Comment:
why 7777.7f ? 
Better leave a comment/note to document the choice

Reply:
The whole block was obsolete, and I deleted it.  (The code was copied from PixelCPETemplateReco, where we handled split clusters.  Here we eventually want to repair clusters by merging split clusters!)

Reply:
The following block is now also commented out, since I realized that we don't need to handle edge pixels separately since on them now we run ClusterRepair by default.